hadoop-ozone/client/src/test/java/org/apache/hadoop/ozone/client/MultiNodePipelineBlockAllocator.java

```diff
@@ -21,68 +21,103 @@
+      int requiredNodes, int clusterSize) {
     this.requiredNodes = requiredNodes;
     this.conf = conf;
+    for (int i = 0; i < clusterSize; i++) {
+      clusterDns.add(HddsProtos.DatanodeDetailsProto.newBuilder().setUuid128(
+          HddsProtos.UUID.newBuilder().setLeastSigBits(i).setMostSigBits(i)
+              .build()).setHostName("localhost").setIpAddress("1.2.3.4")
+          .addPorts(HddsProtos.Port.newBuilder().setName("RATIS").setValue(i)
+              .build()).build());
+    }
+  }
+
+  public List<HddsProtos.DatanodeDetailsProto> getClusterDns(){
+    return this.clusterDns;
   }
 
   @Override
   public Iterable<? extends OzoneManagerProtocolProtos.KeyLocation>
-      allocateBlock(OzoneManagerProtocolProtos.KeyArgs keyArgs) {
-    HddsProtos.Pipeline.Builder builder =
-        HddsProtos.Pipeline.newBuilder().setFactor(keyArgs.getFactor())
-            .setType(keyArgs.getType()).setId(HddsProtos.PipelineID.newBuilder()
-            .setUuid128(HddsProtos.UUID.newBuilder().setLeastSigBits(1L)
-                .setMostSigBits(1L).build()).build());
-    final int rand = RANDOM.nextInt(); // used for port and UUID combination.
-    // It's ok here for port number limit as don't really create any socket
-    // connection.
-    for (int i = 1; i <= requiredNodes; i++) {
-      builder.addMembers(HddsProtos.DatanodeDetailsProto.newBuilder()
-          .setUuid128(HddsProtos.UUID.newBuilder().setLeastSigBits(rand)
-              .setMostSigBits(i).build()).setHostName("localhost")
-          .setIpAddress("1.2.3.4").addPorts(
-              HddsProtos.Port.newBuilder().setName("RATIS").setValue(rand)
-                  .build()).build());
-      if (keyArgs.getType() == HddsProtos.ReplicationType.EC) {
-        builder.addMemberReplicaIndexes(i);
-      }
-    }
-    if (keyArgs.getType() == HddsProtos.ReplicationType.EC) {
-      builder.setEcReplicationConfig(keyArgs.getEcReplicationConfig());
-    }
-    final HddsProtos.Pipeline pipeline = builder.build();
-
+      allocateBlock(OzoneManagerProtocolProtos.KeyArgs keyArgs,
+      ExcludeList excludeList) {
     long blockSize = (long) conf
         .getStorageSize(OzoneConfigKeys.OZONE_SCM_BLOCK_SIZE,
             OzoneConfigKeys.OZONE_SCM_BLOCK_SIZE_DEFAULT, StorageUnit.BYTES);
-
+    long blockGroupLen = keyArgs.getEcReplicationConfig().getData() * blockSize;
+    long dataSize = keyArgs.getDataSize();
     List<OzoneManagerProtocolProtos.KeyLocation> results = new ArrayList<>();
-    results.add(OzoneManagerProtocolProtos.KeyLocation.newBuilder()
-        .setPipeline(pipeline).setBlockID(
-            HddsProtos.BlockID.newBuilder().setBlockCommitSequenceId(1L)
-                .setContainerBlockID(
-                    HddsProtos.ContainerBlockID.newBuilder().setContainerID(1L)
-                        .setLocalID(blockId++).build()).build()).setOffset(0L)
-        .setLength(blockSize).build());
+    long numbBlkGroups = dataSize / blockGroupLen + 1;
```

Comment:
As I understand the code, we are creating multiple block groups here, and return it as preallocated blocks.
If I am right, then I would as if you think it would worth it to allocate less block groupss then then data size, instead of a spare block groups, so that our tests will test the code path where the client has to allocate a new block and can not simply finish the write into pre-allocated ones.

Reply:
I actually cleaned this logic. Currently test are not depending on this pre-creation logic.

Reply:
Alright, we can get back to this later on, as I see now we are just allocating the next actual block now in the multinode mock allocator, I am fine with this approach.